torch/distributed/nn/api/remote_module.py

```diff
@@ -214,105 +209,155 @@ def __init__(
-                    self.on,
-                    _create_module,
-                    (module_cls, args, kwargs, self.device),
-                )
-            else:
-                self.module_rref = module_rref
+            # Create the module on the remote side.
+            self.module_rref = rpc.remote(
+                self.on,
+                _create_module,
+                (module_cls, args, kwargs, self.device),
+            )
 
-        # Install generated methods.
-        for method in self.generated_methods:
-            method_name = method.__name__
-            method = torch.jit.export(method)
-            setattr(self, method_name, types.MethodType(method, self))
+        self._install_generated_methods()
+        self._check_attribute_picklability()
 
-        # Sanity check: whether to be pickled must be explicitly defined for every attribute.
-        for k in self.__dict__.keys():
-            if (
-                k not in _REMOTE_MODULE_PICKLED_ATTRIBUTES
-                and k not in _REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING
-            ):
-                raise AttributeError(
-                    "Attribute {} must be either in ``_REMOTE_MODULE_PICKLED_ATTRIBUTES`` or "
-                    "``_REMOTE_MODULE_ATTRIBUTES_IGNORE_FOR_PICKLING``.".format(k)
-                )
+    def init_from_module_rref(
+        self,
+        remote_device: str,
+        module_rref: rpc.RRef[nn.Module],
+        _module_interface_cls: Any = None,
+    ):
+        """
+        Besides the above constructor, a RemoteModule instance can be initialized given a module RRef,
+        after the user calls ``__new__`` to create an empty instance.
+        This alternate initiailization method can be particularly useful if we want to create multiple
+        RemoteModule instances that share the same underlying module and reduce memory consumption.
+
+        Moreover, this also provides a workaround for passing script RemoteModule over RPC,
+        which is not supported. The recommended way is as follows:
+
+            1. the sender creates a RemoteModule;
+            2. the sender sends its ``module_rref`` over RPC;
+            3. the receiver calls this method to initialize another RemoteModule using the same ``module_rref``.
+
+        Example::
+            Run the following code in two different processes:
+
+            >>> # On worker 0:
+            >>> import torch
+            >>> import torch.distributed.rpc as rpc
+            >>> from torch import nn, Tensor
+            >>> from torch.distributed.nn.api.remote_module import RemoteModule
+            >>>
+            >>> rpc.init_rpc("worker0", rank=0, world_size=2)
+            >>> remote_module = RemoteModule(
+            >>>     "worker1/cpu", nn.Linear, args=(20, 30),
+            >>> )
+            >>>
+            >>> def create_remote_module_by_module_rref(remote_device, module_rref):
+            >>>     remote_module = object.__new__(RemoteModule)
```

Comment:
can we make this `__new__` inside the `init_from_module_rref` instead? i.e. making this function as static function and generate a RemoteModule internally, then return it? A single API should be more intuitive to the user, the UX will be like:

```
remote_module = RemoteModule.init_from_module_rref(remote_device, module_rref)
```

Reply:
Sounds good! Actually this change can also save the work of creating a separate RPC handler.